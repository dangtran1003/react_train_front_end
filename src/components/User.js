import React, { Component} from 'react';
import PropTypes from 'prop-types'
import {user_del} from '../actions/user'
import { connect } from 'react-redux';

class User extends Component {
  static propTypes = {
    username : PropTypes.string,
    name : PropTypes.string,
    email : PropTypes.string,
    id: PropTypes.number
  }

  render() {
    const {username, name, email, id} = this.props
    return (
        <tr>
        <td>{username}</td>
        <td>{name}</td>
        <td>{email}</td>
        <td style={{width: 100}}>
          <a href={"/modify/"+id.toString()} className="btn btn-info" role="button" style={{margin:4}}><i className="icons font-l d-block mt-0 cui-wrench"></i></a>
          <button className="btn btn-danger" value={id} onClick={this.props.user_del}><i className="icons font-l d-block mt-0 cui-trash"></i></button>
        </td>
      </tr>
    )
  }
}

// export default User

const mapDispatchToProps = {
  user_del: user_del,
}

export default connect(null, mapDispatchToProps)(User)