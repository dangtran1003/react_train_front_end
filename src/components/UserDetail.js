import React, { Component} from 'react';
import PropTypes from 'prop-types'
import { connect } from 'react-redux';
import { user_detail, change_detail } from '../actions/user';

export class UserDetail extends Component {
  static propTypes = {
    email : PropTypes.string,
    name : PropTypes.string,
  }
  state = {
      email: '',
      name : '',
      error : ''
  }

  componentWillMount() {
     this.props.UserDetail(this.props.match.params.id)
  }
  handleFormChange = e =>{
       const name = e.target.name
       const value = e.target.value
       this.setState({
        [name] :value
      }) 
   }
   handleSubmit = e =>
        {
            this.props.ChangeDetail(this.props.match.params.id, this.state.email, this.state.name)
        }

  render() {
    return (
        <div className="card">
            <div className="card-header">
                <strong>Cập nhật thông tin nhân viên <i>{this.props.user.username}</i></strong>
            </div>
            <div className="card-body">
                <form className="form-horizontal">
                    <div className="form-group row">
                        <label className="col-md-3 col-form-label" htmlFor="hf-email">Email</label>
                        <div className="col-md-9">
                        <input className="form-control"  type="email" name="email" onChange={this.handleFormChange} autoComplete="email" placeholder={this.props.user.email}/>
                        <span className="help-block">Please enter your email</span>
                        </div>
                    </div>
                    <div className="form-group row">
                        <label className="col-md-3 col-form-label" htmlFor="hf-password">Name</label>
                        <div className="col-md-9">
                            <input className="form-control" type="text" name="name" onChange={this.handleFormChange} placeholder={this.props.user.name}/>
                            <span className="help-block">Please enter your name</span>
                        </div>
                    </div>
                </form>
            </div>
            <div className="card-footer">
                <button className="btn btn-sm btn-primary" onClick={this.handleSubmit}>
                    <i className="fa fa-dot-circle-o"></i> Submit</button>
            </div>
        </div>
    )
  }
}

const mapStateToProps = (state) => ({
    user: state.data.user
  })

function mapDispatchToProps(dispatch) {
    return {
        UserDetail : (id) => {
            dispatch(user_detail(id))
        },
        ChangeDetail : (id, email, name) => {
            dispatch(change_detail(id, email, name))
        }
    }
}
  
export default connect(mapStateToProps, mapDispatchToProps)(UserDetail)