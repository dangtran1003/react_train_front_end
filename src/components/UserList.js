import React, {Component} from 'react'
// import { connect } from "react-redux";
import PropTypes from 'prop-types'
import '../App.css'
import User from './User'
import {user_list, user_add} from '../actions/user'
import { connect } from 'react-redux';

export class UserList extends Component{
    state = {
      username : '',
      email : '',
      name : '',
      error: '', 
      key_search: ''
    }

    handleSubmit = e => {
      // e.preventDefault()
      if (this.state.username.length !== 0 && this.state.email.length !== 0)
        this.props.UserAdd(this.state.username, this.state.email, this.state.name)
      else 
        {
          this.setState({error:'Thiếu username hoặc password!'})
          alert('Thiếu username hoặc password!')
        }
    }

    handleFormChange = e =>{
      const name = e.target.name
      const value = e.target.value
      this.setState({
        [name] :value
      })
      if (name === 'key_search')
      {
        this.props.UserList(value,this.props.match.params.page-1,10)
      }
    }
    static propTypes = {
        users : PropTypes.array,
        total_users: PropTypes.array
    }

    static defaultProps = {
      users : [],
      total_users: [1]
    }

    componentDidMount() {
        this.props.UserList(null, this.props.match.params.page - 1, 10)
    }


    render(){
        return (
          <div className="card">
          <div className="card-header">
            <i> Danh sách người dùng</i>
            </div>
           
            {this.props.error && alert(this.props.error)}
          <div className="card-body">
          <input className="form-control" type="text" name="key_search" onChange={this.handleFormChange} placeholder="Key search"/>
          <br></br>
            <table className="table table-responsive-sm table-bordered table-striped table-sm">
              <thead>
                <tr>
                  <th name='username'>Username</th>
                  <th name='name'>Name</th>
                  <th name='email'>Email</th>
                  <th name='options'>Options</th>
                </tr>
              </thead>
              <tbody>
                {
                   this.props.users.map((user, key) => {
                    const {id, username, email, name} = user
                    return (
                      <User username={username} email={email} name = {name} key={key} id={id}></User>
                    )
                  })
                }
                <tr>
                  <td><input className="form-control" name="username" type="text" placeholder="Enter your username" onChange={this.handleFormChange}></input></td>
                  <td><input className="form-control" name="name" type="text" placeholder="Enter your name" onChange={this.handleFormChange}></input></td>
                  <td><input className="form-control" name="email" placeholder="Enter your email" onChange={this.handleFormChange} type="email"></input></td>
                  <td>
                      <button href="#" className="btn btn-success" style={{margin:4}} onClick={this.handleSubmit}><i className="icons font-l d-block mt-0 cui-check" ></i></button>
                      <button href="#" className="btn btn-danger"><i className="icons font-l d-block mt-0 cui-trash"></i></button>
                  </td>
                </tr>
              </tbody>
            </table>
            <nav>
              <ul className="pagination">
                <li className="page-item">
                  <a className="page-link" href={'list/'+(Number(this.props.match.params.page)>1 ? (Number(this.props.match.params.page)-1).toString() : "") }>Prev</a>
                </li>
                <li className={"page-time" +(this.props.match.params.page === 1 ? " active" : "") }>
                  <a className="page-link" href="list/1">1</a>
                </li>
                {this.props.pages.map((page,key)=>{
                  return (
                    <li className={"page-time" +(this.props.match.params.page === page ? " active" : "") } key = {key}>
                  <a className="page-link" href={'list/' + page.toString()}>{page}</a>
                </li>
                  )
                })}
                <li className="page-item">
                  <a className="page-link" href={'list/'+(Number(this.props.match.params.page) +1).toString()}>Next</a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
        )
    }
}

const mapStateToProps = (state) => ({
  users: state.data.users,
  pages: state.data.pages,
  error: state.data.error
})

function mapDispatchToProps(dispatch) {
  return {
   UserList: (key_search,numPages,numUsers) => {
    dispatch(user_list(key_search, numPages,numUsers))
   },
   
   UserAdd : (username, email, name) => {
     dispatch(user_add(username, email, name))
   }
  }
 }


export default connect(mapStateToProps, mapDispatchToProps)(UserList)