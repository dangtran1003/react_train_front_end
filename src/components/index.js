import {connect} from 'react-redux'
import UserList from './UserList'

export default connect(({users}) => ({users}))(UserList)