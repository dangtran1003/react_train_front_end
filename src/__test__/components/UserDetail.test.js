import {shallow, mount} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import React from 'react'
import {UserDetail}  from '../../components/UserDetail';


const props = { user:
    {
        username : 'trandang',
        name : 'teko vietnam',
        email : 'dang.th@teko.vn',
        id: 1
    },
    match : { params : { page : 1}},
    UserDetail: jest.fn(),
    ChangeDetail: jest.fn(),
}
describe('Components > User', () => {
    //fake event onChange input
    const event = {
        preventDefault() {},
        target: { value: 'the-value' }
      };
    // Snapshot use renderer
    
    it('Render a snapshot for Link use enzyme', () => {
    const tree = shallow(<UserDetail {...props}/>);
    let mockFn = jest.fn()
    tree.instance().handleFormChange = mockFn
        expect(tree.find('div.card-header').text()).toEqual('Cập nhật thông tin nhân viên '+props.user.username)
        //test input form
        expect(tree.find('input').at(0).props().placeholder).toEqual(props.user.email)
        tree.find('input').at(0).simulate('change', event)
        expect(mockFn).toHaveBeenCalledTimes(0)
        expect(tree.find('input').at(1).props().placeholder).toEqual(props.user.name)
        tree.find('input').at(1).simulate('change', event)
        expect(mockFn).toHaveBeenCalledTimes(1)
        //test class button
        expect(tree.find('button').hasClass('btn btn-sm btn-primary')).toEqual(true)
        //test button click
        tree.find('button').simulate('click')
        expect(props.ChangeDetail).toHaveBeenCalledTimes(1)
        expect(shallowToJson(tree)).toMatchSnapshot();
    })
})