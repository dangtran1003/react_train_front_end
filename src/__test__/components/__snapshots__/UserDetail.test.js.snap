// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components > User Render a snapshot for Link use enzyme 1`] = `
<div
  className="card"
>
  <div
    className="card-header"
  >
    <strong>
      Cập nhật thông tin nhân viên 
      <i>
        trandang
      </i>
    </strong>
  </div>
  <div
    className="card-body"
  >
    <form
      className="form-horizontal"
    >
      <div
        className="form-group row"
      >
        <label
          className="col-md-3 col-form-label"
          htmlFor="hf-email"
        >
          Email
        </label>
        <div
          className="col-md-9"
        >
          <input
            autoComplete="email"
            className="form-control"
            name="email"
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [
                    Object {
                      "preventDefault": [Function],
                      "target": Object {
                        "value": "the-value",
                      },
                    },
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            placeholder="dang.th@teko.vn"
            type="email"
          />
          <span
            className="help-block"
          >
            Please enter your email
          </span>
        </div>
      </div>
      <div
        className="form-group row"
      >
        <label
          className="col-md-3 col-form-label"
          htmlFor="hf-password"
        >
          Name
        </label>
        <div
          className="col-md-9"
        >
          <input
            className="form-control"
            name="name"
            onChange={
              [MockFunction] {
                "calls": Array [
                  Array [
                    Object {
                      "preventDefault": [Function],
                      "target": Object {
                        "value": "the-value",
                      },
                    },
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            placeholder="teko vietnam"
            type="text"
          />
          <span
            className="help-block"
          >
            Please enter your name
          </span>
        </div>
      </div>
    </form>
  </div>
  <div
    className="card-footer"
  >
    <button
      className="btn btn-sm btn-primary"
      onClick={[Function]}
    >
      <i
        className="fa fa-dot-circle-o"
      />
       Submit
    </button>
  </div>
</div>
`;
